cou/apps/base.py

```diff
@@ -516,58 +536,74 @@ def _get_upgrade_current_release_packages_step(
 
         return step
 
+    def _get_charm_hub_migration(self) -> PreUpgradeStep:
+        """Get the step for charm hub migration from charm store if necessary.
+
+        :return: Step for charmhub migration
+        :rtype: PreUpgradeStep
+        """
+        if not self.is_from_charm_store:
+            return PreUpgradeStep()
+
+        return PreUpgradeStep(
+            f"Migrate '{self.name}' from charmstore to charmhub",
+            coro=self.model.upgrade_charm(
+                self.name, self.possible_current_channel, switch=f"ch:{self.charm}"
+            ),
+        )
+
+    def _get_change_to_openstack_channels(self) -> PreUpgradeStep:
+        """Get the step for changing to OpenStack channels if necessary.
+
+        :return: Step for changing to OpenStack channels
+        :rtype: PreUpgradeStep
+        """
+        if self.channel != LATEST_STABLE:
+            return PreUpgradeStep()
+        logger.warning(
+            "COU cannot guarantee that it's not a downgrade changing channel of '%s' "
+            "from %s to %s.",
+            self.name,
+            self.channel,
+            self.possible_current_channel,
+        )
+        return PreUpgradeStep(
+            f"WARNING: Changing '{self.name}' channel from {self.channel} to "
+            f"{self.possible_current_channel}. COU cannot guarantee that this is not a downgrade",
+            coro=self.model.upgrade_charm(self.name, self.possible_current_channel),
+        )
+
     def _get_refresh_charm_step(self, target: OpenStackRelease) -> PreUpgradeStep:
         """Get step for refreshing the current channel.
 
-        This function also identifies if charm comes from charmstore and in that case,
-        makes the migration.
         :param target: OpenStack release as target to upgrade.
         :type target: OpenStackRelease
         :raises ApplicationError: When application has unexpected channel.
         :return: Step for refreshing the charm.
         :rtype: PreUpgradeStep
         """
-        if not self.can_upgrade_to:
-            return PreUpgradeStep()
-
-        switch = None
-        *_, channel = self.possible_current_channels
-
-        # corner case for rabbitmq and hacluster.
```

Comment:
where did this corner case end up?

Reply:
Now it's on `possible_current_channel` on the auxiliary.

Reply:
In the end it was just a info-level logline so not a big deal, but I still don't see where that (`len(self.possible_current_channels) > 1`) condition has been relocated